src/main/java/hudson/matrix/MatrixBuild.java

```diff
@@ -350,15 +350,26 @@ protected Result doRun(BuildListener listener) throws Exception {
             MatrixProject p = getProject();
             PrintStream logger = listener.getLogger();
 
-            // give axes a chance to rebuild themselves
-            activeConfigurations = p.rebuildConfigurations(this);
+            try {
+                // lock the project
+                p.buildLock();
+
+                // give axes a chance to rebuild themselves
+                activeConfigurations = p.rebuildConfigurations(this);
+
+                // list up aggregators
+                listUpAggregators(p.getPublishers().values());
+                listUpAggregators(p.getProperties().values());
+                listUpAggregators(p.getBuildWrappers().values());
 
-            // list up aggregators
-            listUpAggregators(p.getPublishers().values());
-            listUpAggregators(p.getProperties().values());
-            listUpAggregators(p.getBuildWrappers().values());
+                // deep copy the axes
+                axes = (AxisList) Jenkins.XSTREAM.fromXML(Jenkins.XSTREAM.toXML(p.getAxes()));
```

Comment:
I do not think this would work because a deep copy is required and the axis do not have a copy constructor or similar (or at least I have not found them)